### PR TITLE
Instance data update closer to notifyDataSetChanged

### LINKED ISSUE #5660: Crash report in 7.0: IndexOutOfBoundsException: Inconsistency detected
```
Fatal Exception: java.lang.IndexOutOfBoundsException: Inconsistency detected. Invalid item position 6(offset:6).state:10
       at android.support.v7.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:5456)
       at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5392)
       at android.support.v7.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:5388)
       at android.support.v7.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2149)
       at android.support.v7.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1533)
       at android.support.v7.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1496)
       at android.support.v7.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:593)
       at android.support.v7.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:3535)
       at android.support.v7.widget.RecyclerView.dispatchLayout(RecyclerView.java:3264)
       at android.support.v7.widget.RecyclerView.onLayout(RecyclerView.java:3796)
       at android.view.View.layout(View.java:16953)
       at android.view.ViewGroup.layout(ViewGroup.java:5579)
       at android.widget.FrameLayout.layoutChildren(FrameLayout.java:396)
       at android.widget.FrameLayout.onLayout(FrameLayout.java:333)
       at android.view.View.layout(View.java:16953)
       at android.view.ViewGroup.layout(ViewGroup.java:5579)
       at android.widget.FrameLayout.layoutChildren(FrameLayout.java:396)
       at android.widget.FrameLayout.onLayout(FrameLayout.java:333)
       at android.view.View.layout(View.java:16953)
       at android.view.ViewGroup.layout(ViewGroup.java:5579)
       at android.support.v7.widget.ActionBarOverlayLayout.onLayout(ActionBarOverlayLayout.java:437)
       at android.view.View.layout(View.java:16953)
       at android.view.ViewGroup.layout(ViewGroup.java:5579)
       at android.widget.FrameLayout.layoutChildren(FrameLayout.java:396)
       at android.widget.FrameLayout.onLayout(FrameLayout.java:333)
       at android.view.View.layout(View.java:16953)
       at android.view.ViewGroup.layout(ViewGroup.java:5579)
       at android.widget.LinearLayout.setChildFrame(LinearLayout.java:2001)
       at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1844)
       at android.widget.LinearLayout.onLayout(LinearLayout.java:1753)
       at android.view.View.layout(View.java:16953)
       at android.view.ViewGroup.layout(ViewGroup.java:5579)
       at android.widget.FrameLayout.layoutChildren(FrameLayout.java:396)
       at android.widget.FrameLayout.onLayout(FrameLayout.java:333)
       at com.android.internal.policy.PhoneWindow$DecorView.onLayout(PhoneWindow.java:2722)
       at android.view.View.layout(View.java:16953)
       at android.view.ViewGroup.layout(ViewGroup.java:5579)
       at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:2552)
       at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2255)
       at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1321)
       at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6708)
       at android.view.Choreographer$CallbackRecord.run(Choreographer.java:894)
       at android.view.Choreographer.doCallbacks(Choreographer.java:696)
       at android.view.Choreographer.doFrame(Choreographer.java:631)
       at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:880)
       at android.os.Handler.handleCallback(Handler.java:815)
       at android.os.Handler.dispatchMessage(Handler.java:104)
       at android.os.Looper.loop(Looper.java:207)
       at android.app.ActivityThread.main(ActivityThread.java:5683)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:789)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:679)
```

### PR DESCRIPTION
**Pull request by hypest**
Fixes #5660 (and possibly #5969 )

The issue cannot reliably be reproduced but based on info gathered (example: https://issuetracker.google.com/issues/37007605) it can be caused when the RecyclerView's adapter reports a wrong item count while its updated.

The fix tries to put the instance data (`mSites`) change closer to the `notifyDataSetChanged()` call so the adapter has consistent data as soon as possible. Previously, `mSites` was changed in the background (`doInBackground`) but the notifyDataSetChanged happened on the main thread (`onPostExecute`).

To test:
* While in "My Site", open the site picker by tapping on "SWITCH SITE"
* Pull to refresh
* The app should normally finish the list update

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java [hunk 1]
```diff
@@ -541,16 +541,22 @@ public int compare(SiteRecord site1, SiteRecord site2) {
...
             }
 
             return null;
         }
 
         @Override
-        protected void onPostExecute(Void results) {
-            notifyDataSetChanged();
+        protected void onPostExecute(SiteList[] updatedSiteLists) {
+            if (updatedSiteLists != null) {
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: I understand the purpose of the change is to make sure we streamline the value change in the class variables to happen as last in time as possible, and issue the `notifyDatSetChanged()` signal right after that. Can we then enclose these 3 below lines in a `synchronized` block to make sure these 3 are packed together in an atomic action?

File: WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java [hunk 2]
```diff
@@ -472,7 +472,7 @@ private SiteList filteredSitesByText(SiteList sites) {
      * AsyncTask which loads sites from database and populates the adapter
      */
     private boolean mIsTaskRunning;
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: I know it's not part of your changes, but what do you think of removing this flag? as it's not used

Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: Sure, cleaned up with cfa672a.

File: WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java (hunk 1 above)
Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: `onPostExecute()` is already executing in main thread (by design) and `notifyDatSetChanged()` will execute in the same thread as well. So, I'd say there is no need for guarding these in a `synchronized` block and will only be "noise" in this case. Would you mind if we leave it as is? Thanks!

**Comment by hypest** [comment]
Thanks for picking this up for review @mzorz ! I've addressed your comments and ready for another pass. Thanks!

File: WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java (hunk 1 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: tl;dr no problem 😁  but let me explain my thoughts: I think the aim here was to make sure changes to the class member vars and the call to notifyDatasetChanged are together as much as possible, but this code, while better than before, doesn't really guarantee that. I suggested using a synchronized block and making sure the only call to `notifyDatasetChanged` is within such block as that would be a good way (probably the only way) to guarantee that.
OTOH it is unusual to have to do this to circumvent a problem in the adapter design itself. We can go with this and see if the problem doesn't appear again

Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: Using a `synchronized` block would be an extreme measure without real value in this case. Problem before was that the instance variable was changed too early (and in a background thread) and the recycler view had the opportunity to repaint its children (for whatever reason) before the `notifyDatasetChanged` call was issued.

Now that the instance variable change and the `notify` both happen on the main thread (and close together) the adapter and the RecyclerView don't have a chance to observe out-of-date data. The `synchronized` block will add no value since there are only 2 code points that `mSites` is changed and both now happen on the main thread. So, `synchronized` would only confuse the reader at this stage.

Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: > without real value in this case.

I agree the value would be marginal, only from a pragmatical perspective.

> don't have a chance to observe out-of-date data

I may be wrong here, but I don't think I agree - the possibility is real, and the only way to _guarantee_ for this not to happen is using a `synchronized` block. Good practices (i.e. changing the dataset and notifying the adapter of a dataset changed, all in the UI Thread, just like your PR here) keep it from happening, but the conditions for it to happen are still probable, so it's perfectly possible (although not likely) to happen. Evidence is here and elsewhere (other apps experiencing the same problem just like what you've already investigated).
Look at the explanation itself: "_close together_": that's a weak condition that can be broken anytime.

> So, synchronized would only confuse the reader at this stage.

A comment explaining things and links to evidence if necessary would get rid of confusions (but this is not what we decided to do here - we decided to merge).

I can keep discussing this but I guess you'll still have something else to say, and I'll as well feel the urge to keep explaining myself,  and we're not adding any value to an already merged PR ¯\_(ツ)_/¯. Let's discuss this over 🍻

Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: > but the conditions for it to happen are still probable

I might still be missing some case so, can you describe a scenario _in this particular Adapter/Fragment/Activity combo_ that the the missing `synchronized` would help avoid an error?

Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: > I might still be missing some case so, can you describe a scenario in this particular Adapter/Fragment/Activity combo that the synchronized would help avoid an error?

The kind of things I can think of atm are better suited for being discussed with beers at this point.

>  it's better to have a specific theory to test and in this case,  the synchronized block is not testing/guarding any specific theory/case.

Theory:
> The fix tries to put the instance data (mSites) change closer to the notifyDataSetChanged() call so the adapter has consistent data as soon as possible.

Basically, the way it looks to me is this current PR is hoping for things to happen, I'm suggesting there is a way to guarantee this will happen.

Comment (hypest) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: > Theory:
>
> > The fix tries to put 

I mean some specific scenario, like perhaps "Background thread <X> may use the updated mSites and because `notifyDataSetChanged()` is not called yet, ...".

I can't envision a scenario like that in this Adapter so far.

Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: me neither

Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/main/SitePickerAdapter.java]: ok been thinking a bit more but couldn't come up with anything else. Let's see how this behaves, and hopefully we can replicate on other adapters as well! 👍
